feat: bound all reconcile loops and network comms with context

devigned opened the pull request:
**What this PR does / why we need it**:

Our controllers have unbounded reconcile loops. That is to say, that the reconcile loop can possibly run forever on a single loop if a stray network call or some other blocking call goes on indefinitely. This PR introduces a command line argument for the controller, `--reconcile-timeout`, which will cause a context timeout if a reconcile loop exceeds the configured time. The default timeout is a generous 10m. It should probably be less than that after async reconciliation is implemented.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #600

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Introduce controller argument `--reconcile-timeout` which provides a max bound on the time a reconcile loop is allowed to execute.
```

CecileRobertMichon commented on util/reconciler/defaults.go:
```diff
@@ -0,0 +1,21 @@
+
+const (
+	// DefaultLoopTimeout is the default timeout for a reconcile loop
+	DefaultLoopTimeout = 10 * time.Minute
```
this might be a little bit optimistic if a single loop creates all the Azure infra / deletes all the azure infra for AzureCluster and AzureMachine.

devigned replied:
Perhaps, you're right. That makes me a little sad 😞. What do you think it should be?

CecileRobertMichon replied:
ARM template timeout is 90 minutes, perhaps that's a good baseline for now? 

PS: Makes me sad too, let's fix it!

devigned commented:
Below are the API changes which bubble up from refactoring to `context.Context`.
```
*** Running go-apidiff ***

sigs.k8s.io/cluster-api-provider-azure/cloud/scope
  Incompatible changes:
  - (*ClusterScope).Close: changed from func() error to func(context.Context) error
  - (*ClusterScope).PatchObject: changed from func() error to func(context.Context) error
  - (*MachinePoolScope).Close: changed from func() error to func(context.Context) error
  - (*MachinePoolScope).GetBootstrapData: changed from func() (string, error) to func(context.Context) (string, error)
  - (*MachinePoolScope).PatchObject: changed from func() error to func(context.Context) error
  - (*MachineScope).Close: changed from func() error to func(context.Context) error
  - (*MachineScope).GetBootstrapData: changed from func() (string, error) to func(context.Context) (string, error)
  - (*MachineScope).PatchObject: changed from func() error to func(context.Context) error
  - ClusterScope.Context: removed

sigs.k8s.io/cluster-api-provider-azure/controllers
  Compatible changes:
  - AzureClusterReconciler.ReconcileTimeout: added
  - AzureMachineReconciler.ReconcileTimeout: added

sigs.k8s.io/cluster-api-provider-azure/exp/controllers
  Compatible changes:
  - AzureMachinePoolReconciler.ReconcileTimeout: added
  - AzureManagedClusterReconciler.ReconcileTimeout: added
  - AzureManagedControlPlaneReconciler.ReconcileTimeout: added
  - AzureManagedMachinePoolReconciler.ReconcileTimeout: added

sigs.k8s.io/cluster-api-provider-azure/util/reconciler
  Compatible changes:
  - DefaultLoopTimeout: added
  - DefaultMappingTimeout: added
  - DefaultedLoopTimeout: added
```
